### ISSUE TITLE
Deep networks barely train: vanishing gradients through the sigmoid, and an initialisation that uses fan-out where it wants fan-in

### ISSUE THREAD

**Issue by xoloki** [xoloki/jlib#131]
Adding hidden layers to `ai::NeuralNetwork` makes training slower without making
it better. That is expected of this architecture, and there are two measurable
reasons, one much larger than the other.

## Measured: the gradient does not reach the early layers

Largest weight change in each layer after one training step, input end first,
64-32-...-32-10 with sigmoid activations:

```
  1 hidden layer    6.42e-03  8.83e-03                                          ratio 0.73
  3 hidden layers   1.38e-03  5.10e-03  8.39e-03  7.79e-03                      ratio 0.18
  6 hidden layers   4.67e-04  1.36e-03  1.77e-03  2.68e-03  5.12e-03  8.69e-03  7.83e-03   ratio 0.06
```

At six hidden layers the layer nearest the input receives **17x less update**
than the layer nearest the loss, and the decay is visible hop by hop at roughly
2-3x per layer.

This is the textbook failure of deep sigmoid networks. Backpropagation
multiplies by the activation's derivative at every hop, and the logistic
sigmoid's derivative is `s(1-s)`, which peaks at 0.25 when `s` is 0.5 and falls
away fast in both directions. Six hops of that is at best 0.25^6, about one
part in four thousand.

So the added depth is parameters that barely move: more arithmetic per epoch,
no more learning. Which is what makes it *look* like the network merely got
slower.

## Compounding it: the initialisation uses the wrong fan

```cpp
T hbound = pow(m_nhidden[0], -0.5);   // m_wih is H x I -- fan-in is I
T obound = pow(m_noutput,    -0.5);   // m_who is O x H -- fan-in is H
T dbound = pow(m_nhidden[i], -0.5);   // m_deep[i] is n[i] x n[i-1] -- fan-in is n[i-1]
```

The `1/sqrt(fan)` form is the right heuristic and the intent is clearly fan-in,
but each one passes the **fan-out**. For a 784-200-10 network:

| matrix | fan-in | bound used | bound wanted | too large by |
|---|---|---|---|---|
| `m_wih` | 784 | 0.0707 | 0.0357 | 2.0x |
| `m_who` | 200 | 0.3162 | 0.0707 | 4.5x |
| `m_deep[i]` | n[i-1] | 1/sqrt(n[i]) | 1/sqrt(n[i-1]) | 1x if the layers are uniform |

Oversized weights make the pre-activations larger, which pushes the sigmoids
toward their saturated ends -- exactly where the derivative is smallest. So the
initialisation makes the vanishing worse rather than being a separate problem,
and `m_who`, the layer the backward pass starts from, is the worst affected.

Note `m_deep` is only wrong when the hidden layers differ in size, which is why
this is easy to miss: the uniform-width case that gets tested most is the one
case it gets right.

## What a fix would need

Two independent changes, in increasing order of disruption:

1. **Fan-in initialisation.** A one-line change per matrix, and it should help
   on its own -- particularly `m_who` at 4.5x. Cheap to try, and measurable
   against a real dataset with `jneural-alpha`.

2. **An activation that does not vanish.** ReLU or leaky ReLU on the hidden
   layers is what made deep networks trainable historically. `m_activation_function`
   is already a `std::function` member, so the forward half is injectable -- but
   the derivative is hardcoded in `train()` as `out ^ (1.0 - out)`, which is
   the sigmoid's and only the sigmoid's. The derivative would have to travel
   with the activation.

Neither should be done blind: the whole claim is about training behaviour, so
it wants a convergence measurement before and after, which means MNIST and
`jneural-alpha` rather than a unit test.

## Relation to #130

#130 is the same family -- an error in the backward pass that compounds with
depth -- but it is roughly **0.1% of one step**, where this is a factor of 17.
Fixing #130 will not make deep networks train. Worth doing them in the other
order, and worth not confusing the two when measuring, since both scale with
depth and either could be credited with an improvement the other produced.

Measured while adding batch support to `train()`; the numbers above come from
comparing `json()` output before and after a single step.

**Comment by xoloki** [comment]
## Fan-in initialisation done and measured; it helps, and it does not fix depth

MNIST, 60k train / 10k test, 64-wide hidden layers, five epochs. Everything in
`jneural-alpha` is default-seeded -- both the weights and the epoch shuffle --
so two runs of a build are identical and the difference between two builds is
the build.

```
                 fan-out (before)   fan-in (after)
  1 x 64             94.21%            94.45%       +0.24
  2 x 64             86.01%            90.63%       +4.62
  3 x 64             10.10%            10.10%        none
```

**Worth 4.6 points at two hidden layers**, which is more than the theory alone
promised, and nothing at three -- where the network sits at 10.1%, which is
chance for ten classes.

### A measurement trap worth recording

The first comparison was at **one epoch**, and it said the opposite:

```
  1 x 64   91.30% -> 92.16%   +0.86
  2 x 64   73.59% -> 71.11%   -2.48     <- fan-in looks worse
```

At one epoch the network is nowhere near converged, so that is a measurement of
the training trajectory rather than of the result, and the ranking flips by five
epochs. Anyone comparing initialisation schemes here should go to five epochs at
least; one epoch will mislead.

## The depth failure is worse than this issue says

The issue says deep networks "barely train". They do not train at all:

```
  hidden layers   accuracy
  1 x 64            94.21%
  2 x 64            86.01%
  3 x 64            10.10%
  4 x 64            10.32%
  6 x 64            10.32%
```

Three or more hidden layers gives exactly chance, and it is a hard floor rather
than slow progress -- 3 x 64 returns 10.32% at one epoch and at three, and at
learning rates 0.01, 0.1 and 0.5. Identical to the digit each time, which is
what a network emitting a constant looks like.

So the cliff is between two and three hidden layers, and initialisation does not
move it. That leaves the activation, which is the second half of this issue and
the part that actually matters: `m_activation_function` is injectable but the
derivative is hardcoded in `train()` as `out ^ (1.0 - out)`, so the two have to
start travelling together before ReLU is possible.

## Also found

`jneural-alpha`'s MNIST path silently inherits the image-mode dimensions.
`INODES = R * C` with defaults R=90, C=120, so without `-r 28 -c 28` it builds a
10800-input network and dies with `mismatch matricies [64,10800], [784,1]` from
`math::matrix` after loading all 60000 rows. Nothing checks the CSV's row length
against the configured input count, and the two modes have no business sharing
a default. Minor, but it is the first thing anyone pointing this at MNIST will
hit.

**Comment by xoloki** [comment]
## Activation is selectable now, and relu beats sigmoid -- but it moves the cliff rather than removing it

`ai::activation` is an enum with the forward function and its derivative kept
together, because the derivative was the half that could not be replaced.
Hidden and output are chosen separately: relu on the output of a classifier
scored against 0.01/0.99 targets kills any unit that goes negative, so hidden
relu with a sigmoid output is the arrangement measured below.

All four derivatives are expressible from the layer's own **output**, which is
what train() already caches -- sigmoid `s(1-s)`, tanh `1-s^2`, relu `s>0?1:0`,
leaky `s>0?1:LEAK`. So nothing needed the pre-activations kept as well.

## Measured

MNIST, 64-wide hidden layers. **The learning rate matters more than the
activation does**, and that nearly buried the result: at the sigmoid default of
0.1, relu is far worse than sigmoid.

```
  1 x 64, relu:   rate 0.1 -> 34.49%   0.01 -> 95.89%   0.003 -> 94.26%   0.001 -> 91.28%
  2 x 64, relu:   rate 0.1 ->  9.80%   0.01 ->  9.80%   0.003 -> 89.92%   0.001 -> 93.03%
```

Sigmoid squashes into [0,1] and relu does not, so the step that suits one
diverges the other. With the rate chosen per configuration:

```
                sigmoid    relu (best rate)
  1 x 64         94.29%    95.89%  (0.01)     +1.6
  2 x 64         87.39%    93.03%  (0.001)    +5.6
  3 x 64         10.10%    88.08%  (0.0003)   from chance to working
  4 x 64         10.32%     9.80%   -- still chance at every rate tried
  6 x 64         10.32%     9.80%   -- likewise
```

**Three hidden layers went from chance to 88%.** So relu does what this issue
predicted, one layer further than sigmoid managed, and the cliff moves from
2->3 rather than disappearing.

## What was ruled out for four layers and beyond

Four is a wall, and it survived everything standard:

- **Learning rate**: swept 0.1, 0.01, 0.003, 0.001, 3e-4, 1e-4, 3e-5, 1e-5.
  Best was 11.35%.
- **Dead units**: leaky_relu tracks plain relu exactly at depth -- 3 layers
  works (87.83% at 3e-4), 4 and 6 do not. If units were dying, leaky would
  rescue them.
- **He initialisation**: sqrt(2/fan) instead of sqrt(1/fan), the standard
  companion to relu. Identical results, including the same 3-layer figure. Not
  the missing piece.
- **Training time**: 4 x 64 at 3e-4 gives 9.8% at both 2 and 10 epochs.

So whatever stops four layers is not vanishing gradients, not saturation, not
initialisation scale and not undertraining. #130 is the obvious remaining
suspect -- it compounds once per layer -- but at ~0.1% per step it looks far too
small to account for a hard floor at chance. Worth a fresh look at the deep
loop in `train()` before assuming it is an optimisation problem.

## A note for whoever measures next

Do not compare at one epoch. An earlier comparison of the fan-in change at one
epoch gave the opposite sign from the same comparison at five, because at one
epoch the network is measuring its trajectory rather than its result. Five
epochs minimum, and the learning rate has to be tuned per activation or the
comparison is between step sizes rather than between activations.

**Comment by xoloki** [comment]
## Closing: both prescriptions landed, and depth trains through five hidden layers now

Re-measured at HEAD today. MNIST 60k/10k, 64-wide hidden layers, five epochs,
default seed -- the same protocol as the comments above, and five epochs
because the trap recorded there is real.

| hidden layers | sigmoid @ 0.1 | relu @ best rate | when this issue was filed |
| --- | --- | --- | --- |
| 1 x 64 | 96.38% | 96.89% (0.01) | 94.21% |
| 2 x 64 | 91.64% | 95.83% (0.001) | 86.01% |
| 3 x 64 | 41.34% | 94.13% (3e-4) | 10.10% -- chance |
| 4 x 64 |  9.58% | **94.36%** (3e-4) | 10.32% -- chance |
| 5 x 64 | -- | **94.05%** (3e-4) | not tried |
| 6 x 64 | 11.35% | 9.80% (six rates swept) | 10.32% -- chance |

**Three, four and five hidden layers went from chance to 94%.** What this issue
asked for is in:

- **fan-in initialisation** (#133) -- worth 4.6 points at two hidden layers,
  and nothing at three, as measured above.
- **an activation that travels with its derivative** (#134) -- which is what
  made relu possible at all, and moved the cliff from 2->3 to 3->4.

And the thing this issue dismissed as too small is what actually broke the
wall: **#130's fix** (#135), taking every gradient at the weights the step
started from. The comment above says "at ~0.1% per step it looks far too small
to account for a hard floor at chance." That was wrong, and it is the most
useful thing on this issue: the magnitude of an error per step said nothing
about whether it was the blocker, because it compounds once per layer and the
failure is a floor rather than a slope.

So the diagnosis in the title -- vanishing gradients through the sigmoid, plus
an initialisation that uses fan-out where it wants fan-in -- is addressed on
both counts, and "deep networks barely train" is no longer true through five
layers.

## What is left is a different question, and is #245

Six hidden layers is still chance at every rate swept (0.01 down to 3e-5). That
survives all three fixes, so it is not this issue's diagnosis and it should not
inherit this issue's title. #245 carries it, with the sweep, what is already
ruled out, and the observation from #141 that makes it interesting: the wall is
being hit by the *more forgiving* of the two error rules, since textbook
backpropagation dies sooner in the same network.

## Two caveats on the numbers above

The sigmoid column is at the default rate only, not swept -- the relu column
uses the best rate per depth. And the rate window at five layers is narrow
(0.001 gives chance, 3e-4 gives 94.05%, 1e-4 gives 71.79%), which is the
strongest form of the warning already on this issue: **the rate matters more
than the activation**, and a depth comparison at one rate is a comparison of
step sizes.
